src/main/kotlin/com/kafkakotlin/demo/users/UserRepository.kt

```diff
@@ -0,0 +1,80 @@
+package com.kafkakotlin.demo.users
+
+import com.kafkakotlin.demo.kafka.producer.KafkaProducer
+import com.kafkakotlin.demo.kafka.statestorequery.StateStoreQuery
+import mu.KLogging
+import org.apache.kafka.streams.StreamsConfig
+import org.apache.kafka.streams.state.KeyValueIterator
+import org.springframework.boot.autoconfigure.kafka.KafkaProperties
+import org.springframework.core.ParameterizedTypeReference
+import org.springframework.http.HttpMethod
+import org.springframework.kafka.config.StreamsBuilderFactoryBean
+import org.springframework.stereotype.Component
+import org.springframework.web.client.RestTemplate
+
+@Component
+class UserRepository(
+    private val kafkaProducer: KafkaProducer,
+    private val store: StateStoreQuery,
+    private val streamsBuilderFactoryBean: StreamsBuilderFactoryBean,
+    private val restTemplate: RestTemplate,
+    kafkaProperties: KafkaProperties
+
+) {
+    private val applicationServer = kafkaProperties.streams.properties[StreamsConfig.APPLICATION_SERVER_CONFIG]?.split(":")
+        ?: throw Exception("Unable to find spring.kafka.properties.application.server property")
+    private val thisHost = applicationServer[0]
+    private val thisPort = applicationServer[1]
+
+    fun createUser(user: User) {
+        return kafkaProducer.strikeMessageToKafka(user)
+    }
+
+    fun getUsers(): Map<String, User> {
+        val metadata = streamsBuilderFactoryBean.kafkaStreams.allMetadataForStore("user-store")
+        val hostAndPortList = metadata.map { data -> mapOf("host" to data.host(), "port" to data.port().toString()) }
+
+        val userList = mutableMapOf<String, User>()
+
+        val localUsers = getLocalUsers()
+        for (user in localUsers) {
+            userList[user.key] = user.value
+        }
+
+        val filteredHosts = hostAndPortList.filterNot { it -> it["host"] == thisHost && it["port"] == thisPort }
+
+        for (entry in filteredHosts) {
+            val remoteUsers = getRemoteUsers(entry["host"], entry["port"])
+            remoteUsers.map { (key, value) -> userList[key] = value }
+        }
+        return userList
+    }
+
+    fun getLocalUsers(): Map<String, User> {
+        return convertKeyValuesToMap(store.getStore().all())
+    }
+
+    fun getRemoteUsers(host: String?, port: String?): Map<String, User> {
```

Comment:
Do the host & port need to be nullable?

Reply:
Also could this be a private method?

Reply:
Happy with the private. On nullable, im getting the host/port values as `entry["host"]` which the compiler is telling me could be nullable. Ideally I could say that part (line 47) should not be nullable, but not sure if that's possible?

Reply:
From your `getRemoteUsers` method point of view, is what would it do if the host or port was null? 

If it doesn't make sense for that to be a valid scenario you should allow it in your method signature.

https://kotlinlang.org/docs/reference/null-safety.html

Reply:
My approach when designing the method like this it to create the public parts of the class first (create the skeleton method) based on how it should be used logically e.g.

```
// host should always be present so make it non nullable
// similar with port
// User doesn't have an ID property so return a Map<user-id, User>
fun getRemoteUsers(host: String, port: String): Map<String, User> {
  // TODO
}
```

Then force the code that calls it to deal with that signature (e.g. ensuring host +port are not null, doing any mapping on the response type).
I find it leads to much more intuitive + readable code in the end

Some people do the above by creating an `interface` first, then implementing it so you think about the code design up front https://medium.com/javarevisited/oop-good-practices-coding-to-the-interface-baea84fd60d3